packetbeat/protos/drda/drda.go

```diff
@@ -0,0 +1,572 @@
+	return stateStrings[state]
+}
+
+func (drda *Drda) getTransaction(k common.HashableTcpTuple) *DrdaTransaction {
+	v := drda.transactions.Get(k)
+	if v != nil {
+		return v.(*DrdaTransaction)
+	}
+	return nil
+}
+
+func init() {
+	protos.Register("drda", New)
+}
+
+func New(
+	testMode bool,
+	results publish.Transactions,
+	cfg *ucfg.Config,
+) (protos.Plugin, error) {
+	p := &Drda{}
+	config := defaultConfig
+	if !testMode {
+		if err := cfg.Unpack(&config); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := p.init(results, &config); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (drda *Drda) init(results publish.Transactions, config *drdaConfig) error {
+	drda.setFromConfig(config)
+
+	drda.transactions = common.NewCache(
+		drda.transactionTimeout,
+		protos.DefaultTransactionHashSize)
+	drda.transactions.StartJanitor(drda.transactionTimeout)
+	//drda.handleMysql = handleMysql
+	drda.results = results
+
+	return nil
+}
+
+func (drda *Drda) setFromConfig(config *drdaConfig) error {
+
+	drda.Ports = config.Ports
+	drda.Send_request = config.SendRequest
+	drda.Send_response = config.SendResponse
+	drda.transactionTimeout = time.Duration(config.TransactionTimeout) * time.Second
+
+	return nil
+
+}
+
+func (drda *Drda) GetPorts() []int {
+	return drda.Ports
+}
+
+func (stream *DrdaStream) PrepareForNewMessage() {
+	stream.data = stream.data[stream.parseOffset:]
+	stream.parseState = drdaStateDDM
+	stream.parseOffset = 0
+	stream.message = nil
+}
+
+//main loop
+//return: ok, complete
+func drdaMessageParser(s *DrdaStream) (bool, bool) {
+
+	m := s.message
+	m.parameters = make(map[uint16]Parameter)
+	for s.parseOffset < len(s.data) {
+
+		direction := ""
+
+		if m.Direction == 0 {
+			direction = "Response"
+		} else {
+			direction = "Request"
+		}
+
+		logp.Debug("drdadetailed", "Direction %s", direction)
+		logp.Debug("drdadetailed", "parser round with parseState = %s and offset: %d, len of data is %d", s.parseState, s.parseOffset, len(s.data))
+
+		switch s.parseState {
+		case drdaStateDDM:
+
+			m.start = s.parseOffset
+			if len(s.data[s.parseOffset:]) < 10 {
+				logp.Err("DRDA DDM Message too short. Ignore it.")
+				return false, false
+			}
+
+			hdr := s.data[s.parseOffset : s.parseOffset+10]
+			if hdr[2] != DRDA_MAGIC {
+				logp.Err("No DRDA magic byte found (%X) but %X", DRDA_MAGIC, uint8(hdr[2]))
+				return false, false
+			}
+
+			if m.ddm.Length != 0 {
+				logp.Err("DDM already initialized.")
+			}
+
+			if m.RemainingLength != 0 {
+				logp.Err("Remaining length must be 0.")
+			}
+
+			ddm := &Ddm{}
+
+			ddm.Length = uint16(hdr[0])<<8 | uint16(hdr[1])
+			ddm.Format = uint8(hdr[3])
+			ddm.DSSType = ddm.Format & 0x0F
+			ddm.DSSFlags = ddm.Format >> 4
+			ddm.Cor = uint16(hdr[4])<<8 | uint16(hdr[5])
+			ddm.Length2 = uint16(hdr[6])<<8 | uint16(hdr[7])
+			ddm.Codepoint = uint16(hdr[8])<<8 | uint16(hdr[9])
+			m.ddm = *ddm
+
+			m.end = int(ddm.Length)
+			m.RemainingLength = int(ddm.Length) - 10
```

Comment:
can we get some sanity checks on Length/RemainingLength in case magic did match for some reason, but we're not really synced to start of some message?